subprojects/base-services/src/main/java/org/gradle/internal/operations/notify/DefaultBuildOperationNotificationListenerRegistrar.java

```diff
@@ -0,0 +1,141 @@
+ * limitations under the License.
+ */
+
+package org.gradle.internal.operations.notify;
+
+import org.gradle.api.Action;
+import org.gradle.internal.progress.BuildOperationDescriptor;
+import org.gradle.internal.progress.BuildOperationListener;
+import org.gradle.internal.progress.OperationFinishEvent;
+import org.gradle.internal.progress.OperationStartEvent;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DefaultBuildOperationNotificationListenerRegistrar implements BuildOperationNotificationListenerRegistrar {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultBuildOperationNotificationListenerRegistrar.class);
+
+    private final Action<? super BuildOperationListener> listenerSubscriber;
+
+    public DefaultBuildOperationNotificationListenerRegistrar(Action<? super BuildOperationListener> listenerSubscriber) {
+        this.listenerSubscriber = listenerSubscriber;
+    }
+
+    @Override
+    public void registerBuildScopeListener(BuildOperationNotificationListener listener) {
+        listenerSubscriber.execute(new Listener(listener));
+    }
+
+    private static class Listener implements BuildOperationListener {
+        private final BuildOperationNotificationListener listener;
+
+        private Listener(BuildOperationNotificationListener listener) {
+            this.listener = listener;
+        }
+
+        @Override
+        public void started(BuildOperationDescriptor buildOperation, OperationStartEvent startEvent) {
+            // replace this with opt-in to exposing on producer side
+            // it just so happens right now that this is a reasonable heuristic
+            if (buildOperation.getDetails() == null) {
+                return;
+            }
+
+            Started notification = new Started(buildOperation.getId(), buildOperation.getParentId(), buildOperation.getDetails());
+            try {
+                listener.started(notification);
+            } catch (Exception e) {
+                LOGGER.debug("Build operation notification listener threw an error on " + notification, e);
+            }
+        }
+
+        @Override
+        public void finished(BuildOperationDescriptor buildOperation, OperationFinishEvent finishEvent) {
+            // replace this with opt-in to exposing on producer side
+            // it just so happens right now that this is a reasonable heuristic
+            if (finishEvent.getResult() == null) {
+                return;
+            }
+
+            Finished notification = new Finished(buildOperation.getId(), finishEvent.getResult());
+            try {
+                listener.finished(notification);
+            } catch (Exception e) {
+                LOGGER.debug("Build operation notification listener threw an error on " + notification, e);
+            }
+        }
+    }
+
+    private static class Started implements BuildOperationStartedNotification {
```

Comment:
this is an optimisation, but to avoid a bunch of garbage we might have the original events implement these interfaces instead of adapting.